### ISSUE TITLE
Make `Object.hashAll` and `Object.hashAllUnordered` accept nullable lists

### ISSUE THREAD

**Issue by werainkhatri** [dart-lang/sdk#48094]
Unlike `Object.hash`, `Object.hashAll` and `Object.hashAllUnordered` does not accept nullable parameters, which it should to make it consistent with `Object.hash` and flutter's (soon-to-be previous) hash function (`hashList`).

**Comment by lrhn** [comment]
I disagree with doing this change.

It's not consistent with `Object.hash`. The `Object.hash` function accepts a single object, which may be `null` (because `null` is an object).
The `Object.hashAll` accepts zero or more objects, as a list. The list elements can be `null`. Making the list be `null` does not match that. it's unnecessary.


There is no advantage in allowing you to pass in `null`, because you can just not call the function at all if you have a `null` value.
If you have a potentially `null` list value, and want to hash it, just do `Object.hashAll(maybeList ?? const [])`, or make a helper function:

```dart
int myHashAll(List<Object?>? maybeList) => maybeList == null ? null.hashCode : Object.hashAll(maybeList);
```
You can then use that *if you need it*, but not if you have a non-`null` list.

If we make `Object.hashAll` accept `null` as argument, even though it doesn't *need* it, that can hide mistakes in the program, where a value is `null` unexpectedly. You won't get any help discovering such a bug. 
Example:
```dart
int something(Object x) {
  List<Object?>? collection;
  if (x is SpecialThing) {
    // Always true, I almost promise!
    collection = x.listOfStuff;
  }
  // some more code here.
  return Object.hashAll(collection);
}
```
This code will get a warning now, because `collection` is potentially `null`.
If `Object.hashAll` accepted `null`, you wouldn't get that warning.

So, I'm firmly against allowing `hashAll` to accept `null`. It's bad design.
Even if we had introduced the functions prior to Null Safety, I'd still have made them throw on a `null` argument.

**Comment by Hixie** [comment]
@lrhn I would be curious to have you review the PR above that migrates Flutter to this new API. There's a number of places where we're being forced to add `?? <Object>[]` which seems really quite unfortunate and not at all obvious, and I wonder if you think they represent places where we've made poor API design decisions that we should rethink.

**Comment by Hixie** [comment]
(If you have any advice on how we should implement the `dart fix` given the API difference that would be useful too.)

**Comment by lrhn** [comment]
If it was me, I'd probably use `?? const<Object>[]` to avoid allocation (but risk polymorphism, at least if `Object.hashAll` is inlined - if not, it's almost certainly polymorphic already), or I'd use `c == null ? 3513 : Object.hashAll(c)` if the expression is a variable already.
 
Not allowing `null` makes it slightly harder to use when you have either no list or a (probably potentially empty) list, but that itself is something I'd try to avoid too. Is there a logical difference between no list and an empty list? 
If not, don't have two states for the same thing.
If yes, you probably want *different* hash codes for the two states.

The meaning of `null` in this context is part of the caller's domain logic, the `Object.hashAll` function cannot know what it means, it can only choose to ignore it. If anything, it should return `null` on a `null` argument, so you can see it happened, and handle it. That would make the return type `int?` which would annoy every other user of the function. Better to stick to the one job it has: Hash a sequence (iterable) of values.

Not allowing `null` is safer than allowing `null` for an operation which doesn't *need* to handle `null`. It won't hide accidental uses of nullable variables. It avoids assigning a specific meaning to a value outside of the domain that the `hashAll` function is actually defined on.

**Comment by Hixie** [comment]
Fair enough. I think that's more theoretically right than pragmatically right but I can respect the reasoning.
